[Global] 글로벌 예외 핸들러 구현

codingbaraGo opened the pull request:
<!-- 제목은 `[도메인] PR 제목`으로 작성해주세요. (ex) [User] 소셜 로그인 필터 리팩토링 -->
<!-- PR의 타입은 Label을 통해 나타내주세요. -->

### 💻 작업 내용
<!-- 진행한 작업 내용에 대해 작성해주세요. -->
- ErrorCode 인터페이스 정의
- Error 응답 객체 정의
- Exception handler 구현
- 공통 Exception 구현 

### ✨ 리뷰 포인트
<!-- (ex) query 가 너무 복잡한 것 같은데 이 위주로 봐주세요. -->
- 커스텀 exception 객체와 ErrorCode의 형식이 괜찮은지 위주로 봐주세요.

### 📝 메모
<!-- PR 관련 전달사항이 있다면 이곳에 작성해주세요. -->
글로벌 예외 핸들러는 실행 모듈인 module-app에 넣었습니다.
첫 PR인 만큼 모듈의 형태도 한번 봐주시면 감사하겠습니다.

### 🎯 관련 이슈
<!-- pr이 merge 되면 이슈가 자동으로 close 되도록 합니다. 만약 자동 close 를 하지 않고 이슈만 링크한다면 closed 키워드를 삭제해주세요.-->
closed #2

codingbaraGo committed:
feat(global): Global exception handler 구현
- ErrorCode 인터페이스 정의
- Error 응답 객체 정의
- Exception handler 구현
- 공통 Exception 구현

devdnwls commented on module-app/src/main/java/com/mema/server/global/exception/ErrorCode.java:
```diff
@@ -0,0 +1,40 @@
+ */
+@Getter
+@JsonFormat(shape = JsonFormat.Shape.OBJECT)
+public enum ErrorCode {
```
ErrorCode를 인터페이스로 정의하셨다고 PR에 작성해주셨는데 잘못 작성하신 것이 맞나요?

codingbaraGo replied:
인터페이스를 정의해 추가했습니다!
잘못 작성한것이었는데, 인터페이스를 사용하는편이 각 도메인에서 의존성 없이 에러 관련 수정을 하기 유리하겠더라구요.

devdnwls commented on module-app/src/main/java/com/mema/server/global/exception/ErrorCode.java:
```diff
@@ -0,0 +1,40 @@
+/**
+ * <b>Error Code Enum</b>
+ *
+ */
```
Java Doc 부분 주석은 클래스 네이밍이 명확하다면 필요없을 것 같아요.

devdnwls commented on module-app/src/main/java/com/mema/server/global/exception/ErrorCode.java:
```diff
@@ -0,0 +1,40 @@
+
+    /* 5xx - Server Error */
+    INTERNAL_ERROR(HttpStatus.INTERNAL_SERVER_ERROR, "500_INTERNAL", "서버 내부 오류가 발생했습니다."),
+    DATABASE_ERROR(HttpStatus.INTERNAL_SERVER_ERROR, "500_DATABASE", "데이터베이스 처리 중 오류가 발생했습니다."),
```
서버 측의 예외라면 하나로 충분히 묶을 수 있을 것 같아요.
`MEMA_SERVER_EXCEPTION` 이런 식으로요. 클라이언트 측도 마찬가지입니다.

codingbaraGo replied:
클라이언트 측 에러를 한번에 묶어버린다면, 클라이언트에서 에러에 대한 처리를 하기 힘들어지지 않을까요?
일단 최대한 줄여놨습니다.

devdnwls replied:
답변을 드렸으니 고민해보시면 좋을 것 같아요

devdnwls commented on module-app/src/main/java/com/mema/server/global/exception/ErrorCode.java:
```diff
@@ -0,0 +1,40 @@
+    /* 4xx - Client Error */
+    INVALID_INPUT(HttpStatus.BAD_REQUEST, "400_INVALID_INPUT", "입력 값이 올바르지 않습니다."),
+    MISSING_PARAMETER(HttpStatus.BAD_REQUEST, "400_MISSING_PARAM", "필수 파라미터가 누락되었습니다."),
+    VALIDATION_FAILED(HttpStatus.BAD_REQUEST, "400_VALIDATION_FAIL", "유효성 검증에 실패했습니다."),
```
파라미터와 관련된 내용이니까 얘도 하나로 합칠 수 있을 것 같아요.

codingbaraGo replied:
반옇했습니다.

devdnwls commented on module-app/src/main/java/com/mema/server/global/exception/ErrorCode.java:
```diff
@@ -0,0 +1,40 @@
+    MISSING_PARAMETER(HttpStatus.BAD_REQUEST, "400_MISSING_PARAM", "필수 파라미터가 누락되었습니다."),
+    VALIDATION_FAILED(HttpStatus.BAD_REQUEST, "400_VALIDATION_FAIL", "유효성 검증에 실패했습니다."),
+
+    API_NOT_FOUND(HttpStatus.NOT_FOUND, "404_API_NOT_FOUND", "요청한 API를 찾을 수 없습니다."),
```
요청한 API가 없는 의도라는 것은 명확합니다만 사실 이것도 클라이언트 측에 포함시키면 좋을 것 같아요.

devdnwls commented on module-app/src/main/java/com/mema/server/global/exception/ErrorCode.java:
```diff
@@ -0,0 +1,40 @@
+    VALIDATION_FAILED(HttpStatus.BAD_REQUEST, "400_VALIDATION_FAIL", "유효성 검증에 실패했습니다."),
+
+    API_NOT_FOUND(HttpStatus.NOT_FOUND, "404_API_NOT_FOUND", "요청한 API를 찾을 수 없습니다."),
+    NO_SUCH_RESOURCE(HttpStatus.NOT_FOUND, "404_RESOURCE_NOT_FOUND", "요청한 리소스를 찾을 수 없습니다."), // <-- 코드값 수정
```
RESOURCE_NOT_FOUND가 적절할 것 같아요.

devdnwls replied:
추가로 이미 상태코드에서 404가 드러나는데 설명 측에 404를 붙일 필요는 없을 것 같아요

devdnwls commented on module-app/src/main/java/com/mema/server/global/exception/ErrorCode.java:
인터페이스를 정의한 것이라면 필요한 메서드만을 두고 각 도메인별(예를 든다면 인증이면 AuthenticationErrorCode)로 ENUM 타입을 만들어 구현하면 유지보수도 용이할 거고 명확하게 어느 쪽에 위치하는지가 보일 것 같아요.

codingbaraGo replied:
인터페이스를 정의해 각 도메인별로 ENUM 타입을 만들어 정의하도록 수정하였습니다.

devdnwls commented on module-app/src/main/java/com/mema/server/global/exception/ErrorResponse.java:
```diff
@@ -0,0 +1,51 @@
+/**
+ * <b>에러 응답 DTO</b>
+ *
+ */
```
마찬가지

devdnwls commented on module-app/src/main/java/com/mema/server/global/exception/ErrorResponse.java:
```diff
@@ -0,0 +1,51 @@
+        int status,
+        String path,
+        LocalDateTime timestamp,
+        @JsonInclude(JsonInclude.Include.NON_EMPTY) List<FieldError> fieldErrors) {
```
Field에 대한 검증을 내보내주는 것은 사실 좀 고민을 많이 해야해요.
노출시키고 싶지 않은 예외를 그대로 내보내주게 된다면 사이트의 보안이라든가 허점이 잡힐 수 있는 내용이 있기 때문이거든요.

BindingResult로 각 필드를 꼼꼼하게 검증하려는 의도로 보이시는데 조금 고민을 더 해보시는게 좋을 것 같아요.

codingbaraGo replied:
global 핸들러에서는 그냥 에러 코드만 반환하게 했습니다.
필요하다면 각 도메인에서 필요한 부분에 대해서만 검증 에러 정보를 반환하면 되겠군요!

devdnwls commented on module-app/src/main/java/com/mema/server/global/exception/ErrorResponse.java:
```diff
@@ -0,0 +1,51 @@
+     *
+     * @param field  오류가 난 필드명
+     * @param reason 오류 메시지
+     */
```
마찬가지

devdnwls commented on module-app/src/main/java/com/mema/server/global/exception/ErrorResponse.java:
```diff
@@ -0,0 +1,51 @@
+                req.getRequestURI(),
+                LocalDateTime.now(),
+                errors);
+    }
```
Indentation은 보기 좋게 바꾸시는 것이 좋을 것 같아요.

codingbaraGo replied:
저는 좌측정렬이 되어있어 깔끔하다 생각했는데,
어떤식으로 바꾸는게 좋을까요?

devdnwls replied:
파라미터가 많아서 한 줄로 길게 이어지면 옆으로 넘겨서 봐야하는 수고로움을 덜기 위해 원라인씩 끊어 쓰시는게 가독성은 더 좋을거에요

devdnwls commented on module-app/src/main/java/com/mema/server/global/exception/GlobalExceptionHandler.java:
```diff
@@ -0,0 +1,110 @@
+public class GlobalExceptionHandler {
+
+
+    private void logEx(HttpServletRequest req, Exception e) {
+        log.error(
+                "[EXCEPTION] uri={} method={} type={} msg={}",
+                req.getRequestURI(),
+                req.getMethod(),
+                e.getClass().getSimpleName(),
+                e.getMessage(),
+                e);
+    }
```
로깅을 위한 것으로 보여지는데 별도의 유틸리티 클래스에서 점진적으로 추가할 수 있도록 하는 것이 좋을 것 같아요
컨트롤러 전역 핸들러에서의 책임이라고 보기엔 어려울 것 같아요.

codingbaraGo replied:
넵
일단 지금은 분리하지 않았습니다.
나중에 따로 분리해 로깅 관련 로직을 작성해야겠군요.

devdnwls commented on module-app/src/main/java/com/mema/server/global/exception/GlobalExceptionHandler.java:
```diff
@@ -0,0 +1,110 @@
+    public ResponseEntity<ErrorResponse> handleService(ServiceException e, HttpServletRequest req) {
+        logEx(req, e);
+        return build(e.getErrorCode(), req);
+    }
```
너무 추상적이에요.
각 도메인마다 공통 예외 응답을 주는 경우는 극히 드뭅니다.
게다가 만약 특정 도메인에서 이러이러한 내용을 추가해달라고 한다면 결합도가 높아서 다른 곳에까지도 영향을 미칠 것 같아요.

devdnwls commented on module-app/src/main/java/com/mema/server/global/exception/ServiceException.java:
```diff
@@ -0,0 +1,30 @@
+        this.errorCode = errorCode;
+    }
+
+    public static ServiceException of(ErrorCode code) {
```
정적 팩토리 메서드의 네이밍에 대한 규칙을 공부해보시는게 좋을 것 같아요

codingbaraGo committed:
refactor(Global): ErrorCode 인터페이스 정의 (#2)
- ErrorCode 인터페이스 정의
- 기존 ErrorCode의 공통 에러 코드들을 CommonErrorCode 클래스로 이동
- 공통 에러 코드 축소

codingbaraGo committed:
refactor(Global): 공통 Exception abstract class 정의 (#2)
- 기존 ServiceException 클래스 삭제
- 공통 Exception 클래스 정의
- default Exception class 정의

codingbaraGo committed:
refactor(Global): Error 응답 형식 수정 (#2)
- ErrorResponse 인터페이스 정의
- Binding result 관련 코드 삭제
- default 클래스 정의

codingbaraGo committed:
refactor(Global): GlobalExceptionHandler 수정 (#2)
- Binding result 관련 코드 삭제
- 도메인 Exception은 각 도메인에서 처리하고, Global handler에서는 공통 에러 및 unhandled 에러만 처리하도록 수정

codingbaraGo committed:
fix(Global): BaseException 의존성 수정 (#2)
- BaseException 클래스에서 ErrorCode가 아닌, 구현체에 의존하는 문제를 수정

codingbaraGo committed:
refactor(Global): 정적 팩토리 메서드 명 수정  (#2)
